### PR TITLE
For science

### PR DESCRIPTION
**Pull request by jdayton3**
## Description

Add the For Science keyboard to QMK

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/for_science/for_science.c [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    // runs for every action, just before processing by the firmware
+
+    return process_record_user(keycode, record);
+}
+
+void led_set_kb(uint8_t usb_led) {
+    // put your keyboard LED indicator (ex: Caps Lock LED) toggling code here
+
+    led_set_user(usb_led);
+}
```
Comment (zvecr) [inline keyboards/for_science/for_science.c:48]:
```suggestion
```

File: keyboards/for_science/config.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+#define DIODE_DIRECTION ROW2COL
+
+/* Debounce reduces chatter (unintended double-presses) - set 0 if debouncing is not needed */
+#define DEBOUNCE 5
+
+#define IGNORE_MOD_TAP_INTERRUPT
+#define TAPPING_FORCE_HOLD
+#define PERMISSIVE_HOLD
+#define PREVENT_STUCK_MODIFIERS
+
```
Comment (zvecr) [inline keyboards/for_science/config.h:59]:
```suggestion

```

File: keyboards/for_science/config.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+//#define MASTER_RIGHT
+
+/* key combination for magic key command */
+#define IS_COMMAND() ( \
+    keyboard_report->mods == (MOD_BIT(KC_LSHIFT) | MOD_BIT(KC_RSHIFT)) \
+)
```
Comment (zvecr) [inline keyboards/for_science/config.h:65]:
```suggestion
/* key combination for magic key command */
/* defined by default; to change, uncomment and set to the combination you want */
// #define IS_COMMAND() (get_mods() == MOD_MASK_SHIFT)
```

**Comment by jdayton3** [comment]
@zvecr I added in all the changes you requested, plus an info.json

File: keyboards/for_science/rules.mk [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
+MIDI_ENABLE = no            # MIDI support (+2400 to 4200, depending on config)
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
```
Comment (fauxpark) [inline keyboards/for_science/rules.mk:22]:
```suggestion
#
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = yes        # Console for debug
COMMAND_ENABLE = yes        # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
MIDI_ENABLE = no            # MIDI support
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

**Comment by TobyGWilliams** [comment]
Any news on this PR? I'd love to get my keyboard programmed

File: keyboards/for_science/keymaps/default/keymap.c [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        if (record->event.pressed) {
+            layer_on(_MOD_LAYER);
+        } else {
+            layer_off(_MOD_LAYER);
+        }
+    }
+
+    if (record->event.pressed) {
+        if (keycode == KC_BSPC && keyboard_report->mods & MODS_ALT) {
+            SEND_STRING(SS_TAP(X_DELETE));
```
Comment (drashna) [inline keyboards/for_science/keymaps/default/keymap.c:113]: IIRC, this compiles smaller: 
```suggestion
           tap_code(KC_DEL);
```

If you're using longer strings, then the `SEND_STRING` function is better.

**Comment by setek** [comment]
Hi, I have made the changes as requested. Please let me know if there’s anything else you need me to do 😄

File: keyboards/for_science/for_science.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include "quantum.h"
+
+#define LAYOUT( \
```
Comment (noroadsleft) [inline keyboards/for_science/for_science.h:20]:
```suggestion
#define LAYOUT_split_4x5_3( \
```

File: keyboards/for_science/info.json [hunk 7]
```diff
@@ -0,0 +1,197 @@
+{
+    "keyboard_name": "For Science",
+    "url": "https://github.com/peej/for-science-keyboard",
+    "maintainer": "qmk",
+    "width": 11.25,
+    "height": 5,
+    "layouts": {
+        "LAYOUT": {
```
Comment (noroadsleft) [inline keyboards/for_science/info.json:8]:
```suggestion
        "LAYOUT_split_4x5_3": {
```

File: keyboards/for_science/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+ * |-Shift-+-------+-------+-------+-------+   |-------+-------+-------+-------+-------|
+ * |   Z   |   X   |   C   |   V   |   B   |   |   N   |   M   |   ,   |   .   |   /   |
+ * |  F6   |  F7   |  F8   |  F9   |  F10  |   | PgDn  | Left  | Down  | Right |       |
+ * `-Shift---Layer-+-------+-------+-------+   |-------+-------+-------+---------Layer-'
+ *                 | Ctrl  |  Gui  |  Alt  |   | Space | Layer | Shift |
+ *                 |       |       |       |   | Funct |       |       |
+ *                 `------------------------   `-Layer-----------------'
+ */
+
+[_BASE] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/for_science/keymaps/default/keymap.c:62]:
```suggestion
[_BASE] = LAYOUT_split_4x5_3(
```

File: keyboards/for_science/keymaps/default/keymap.c [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    KC_TAB,  KC_ESC,  KC_LPRN, KC_LCBR, KC_LBRC,    KC_RBRC, KC_RCBR, KC_RPRN, KC_QUOT, KC_BSPC,
+    KC_Q,    KC_W,    KC_E,    KC_R,    KC_T,       KC_Y,    KC_U,    KC_I,    KC_O,    KC_P,
+    SFT_A,   KC_S,    KC_D,    KC_F,    KC_G,       KC_H,    KC_J,    KC_K,    KC_L,    KC_ENT,
+    SFT_Z,   LAY_X,   KC_C,    KC_V,    KC_B,       KC_N,    KC_M,    KC_COMM, KC_DOT,  LAY_SLS,
+                      KC_LCTL, KC_LGUI, KC_LALT,    LAY_SPC, LAYER,   KC_RSFT
+),
+
+[_LAYER] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/for_science/keymaps/default/keymap.c:70]:
```suggestion
[_LAYER] = LAYOUT_split_4x5_3(
```

File: keyboards/for_science/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+    KC_TILD, KC_GRV,  KC_LABK, KC_UNDS, KC_MINS,    KC_PLUS, KC_EQL,  KC_RABK, KC_BSLS, KC_DEL,
+    KC_1,    KC_2,    KC_3,    KC_4,    KC_5,       KC_6,    KC_7,    KC_8,    KC_9,    KC_0,
+    SFT_F1,  KC_F2,   KC_F3,   KC_F4,   KC_F5,      KC_PGUP, KC_HOME, KC_UP,   KC_END,  KC_SCLN,
+    SFT_F6,  LAY_F7,  KC_F8,   KC_F9,   KC_F10,     KC_PGDN, KC_LEFT, KC_DOWN, KC_RGHT, _______,
+                      _______, _______, _______,    FUNCT,   _______, _______
+),
+
+[_MOD_LAYER] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/for_science/keymaps/default/keymap.c:78]:
```suggestion
[_MOD_LAYER] = LAYOUT_split_4x5_3(
```

File: keyboards/for_science/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    KC_TILD, _______, _______, _______, _______,    KC_Q,    _______, _______, _______, _______,
+    KC_TAB,  _______, _______, _______, _______,    _______, _______, _______, _______, _______,
+    _______, _______, _______, _______, _______,    _______, _______, _______, _______, _______,
+    _______, _______, _______, _______, _______,    _______, _______, _______, _______, _______,
+                      _______, _______, _______,    _______, _______, _______
+),
+
+[_FUNCT] = LAYOUT(
```
Comment (noroadsleft) [inline keyboards/for_science/keymaps/default/keymap.c:86]:
```suggestion
[_FUNCT] = LAYOUT_split_4x5_3(
```

File: keyboards/for_science/rules.mk [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = yes        # Console for debug
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
```
Comment (noroadsleft) [inline keyboards/for_science/rules.mk:20]:
```suggestion
```

File: keyboards/for_science/config.h [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xFEED
+#define PRODUCT_ID      0x0001
```
Comment (noroadsleft) [inline keyboards/for_science/config.h:21]:
```suggestion
#define PRODUCT_ID      0xF22C
```

File: keyboards/for_science/config.h [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    Peej
+#define PRODUCT         For Science
+#define DESCRIPTION     A 100x100mm 5x4+3 split keyboard
```
Comment (noroadsleft) [inline keyboards/for_science/config.h:25]: `DESCRIPTION` is deprecated and no longer used anywhere within QMK.

```suggestion
```

**Comment by setek** [comment]
> GitHub tip: You can apply multiple suggestions to a single commit by using the [Files Changed tab](https://github.com/qmk/qmk_firmware/pull/9607/files).

wow thanks for that, I had no idea 😅 so easy!

I’m not a regular github fork merger kinda guy. We use bitbucket at work

**Review by fauxpark — APPROVED** [review]

**Comment by noroadsleft** [comment]
Thanks!

**Pull request merged**